File: package/main/src/Color/cmykToRgba.ts
import { clamp } from "@/Math/clamp";
import { division } from "@/Math/division";
import { multiplication } from "@/Math/multiplication";
import { roundOf } from "@/Math/roundOf";
import { subtract } from "@/Math/subtract";

/**
 * Convert CMYK color values to RGBA color space
 * @param {number} c Cyan percentage (0-100)
 * @param {number} m Magenta percentage (0-100)
 * @param {number} y Yellow percentage (0-100)
 * @param {number} k Key/Black percentage (0-100)
 * @param {number} a Alpha value (0-1)
 * @returns {Object} RGBA values (r, g, b as 0-255, a as 0-1)
 * @example cmykToRgba(100, 100, 0, 60.78) // { r: 0, g: 0, b: 100, a: 1 }
 */
export const cmykToRgba = (
  c: number,
  m: number,
  y: number,
  k: number,
  a = 1,
): { r: number; g: number; b: number; a: number } => {
  // Clamp CMYK values to 0-100 range
  const clampedC = clamp(c, 0, 100);
  const clampedM = clamp(m, 0, 100);
  const clampedY = clamp(y, 0, 100);
  const clampedK = clamp(k, 0, 100);

  // Convert CMYK values to 0-1 range
  const cPercentage = division(clampedC, 100);
  const mPercentage = division(clampedM, 100);
  const yPercentage = division(clampedY, 100);
  const kPercentage = division(clampedK, 100);

  // Calculate RGB values
  const r = multiplication(
    255,
    subtract(1, cPercentage),
    subtract(1, kPercentage),
  );
  const g = multiplication(
    255,
    subtract(1, mPercentage),
    subtract(1, kPercentage),
  );
  const b = multiplication(
    255,
    subtract(1, yPercentage),
    subtract(1, kPercentage),
  );

  // Round RGB values to 0-255 range
  const roundedR = roundOf(r);
  const roundedG = roundOf(g);
  const roundedB = roundOf(b);

  // Clamp alpha value to 0-1 range
  const clampedA = clamp(a, 0, 1);

  return {
    r: roundedR,
    g: roundedG,
    b: roundedB,
    a: clampedA,
  };
};

File: package/umt_python/src/color/cmyk_to_rgba.py
def cmyk_to_rgba(
    c: float, m: float, y: float, k: float, a: float = 1.0
) -> dict[str, float]:
    """
    Convert CMYK color values to RGBA color space.

    Args:
        c: Cyan percentage (0-100)
        m: Magenta percentage (0-100)
        y: Yellow percentage (0-100)
        k: Key/Black percentage (0-100)
        a: Alpha value (0-1)

    Returns:
        Dictionary with r, g, b as 0-255 and a as 0-1

    Example:
        >>> cmyk_to_rgba(100, 100, 0, 60.78)
        {'r': 0, 'g': 0, 'b': 100, 'a': 1.0}
    """
    clamped_c = max(0, min(100, c))
    clamped_m = max(0, min(100, m))
    clamped_y = max(0, min(100, y))
    clamped_k = max(0, min(100, k))

    c_percentage = clamped_c / 100
    m_percentage = clamped_m / 100
    y_percentage = clamped_y / 100
    k_percentage = clamped_k / 100

    r = 255 * (1 - c_percentage) * (1 - k_percentage)
    g = 255 * (1 - m_percentage) * (1 - k_percentage)
    b = 255 * (1 - y_percentage) * (1 - k_percentage)

    rounded_r = round(r)
    rounded_g = round(g)
    rounded_b = round(b)

    clamped_a = max(0, min(1, a))

    return {"r": rounded_r, "g": rounded_g, "b": rounded_b, "a": clamped_a}
